controllers/sync_namespace.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2023 KubeAGI.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/minio/minio-go/v7"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/kubeagi/arcadia/pkg/config"
+	"github.com/kubeagi/arcadia/pkg/datasource"
+	"github.com/kubeagi/arcadia/pkg/utils"
+)
+
+const (
+	BucketNotEmpty         = "The bucket you tried to delete is not empty"
+	BucketNotExist         = "The specified bucket does not exist"
+	SkipNamespaceConfigMap = "skip-namespaces"
+)
+
+type NamespacetReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Dataset object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.2/pkg/reconcile
+func (r *NamespacetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	var err error
+	instance := &v1.Namespace{}
+	if err = r.Client.Get(ctx, req.NamespacedName, instance); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	skip, err := r.checkSkippedNamespace(ctx, instance.Name)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if skip {
+		klog.Infof("namespace %s is in the filter list and will not be created, delete the corresponding bucket.", instance.Name)
+		return reconcile.Result{}, nil
+	}
+
+	if instance.DeletionTimestamp != nil {
+		err = r.removeBucket(ctx, instance.Name)
+		return reconcile.Result{RequeueAfter: 3 * time.Second}, err
+	}
+
+	err = r.syncBucket(ctx, instance.Name)
```

Comment:
so all events from namespace will invoke syncBucket method? is there any attribute( annotation etc..) that can be used to limit the rate of syncBucket, it's better to invoke once when a specific event occurs.

Reply:
Maybe the `create` and `delete` actions do the processing just fine, and the `update` operation, can just be ignored.